File: api/_lib/safety-engine.js
import { CONFIG } from './config.js';

/**
 * Evaluates multi-sensor telemetry according to safety thresholds and physics-based corroboration.
 * VitalLume strictly provides ambient environmental safety and hazard detection research classifications.
 * It DOES NOT provide medical diagnoses.
 * 
 * Classifications:
 * - NORMAL
 * - WARNING
 * - CRITICAL
 * - UNKNOWN
 */
export function evaluateSafetyConditions(data) {
  const alerts = [];
  const flags = [];
  let highestSeverity = 'NORMAL';

  const temp = data.temperature != null ? Number(data.temperature) : null;
  const humidity = data.humidity != null ? Number(data.humidity) : null;
  const gas = data.gas != null ? Number(data.gas) : null;
  const sound = data.sound != null ? Number(data.sound) : null;
  const motion = Boolean(data.motion || data.pir);
  const presence = Boolean(data.presence);
  
  // Radar kinematics
  const radar = data.radar || {};
  const radarMotion = Boolean(radar.motion || data.motion);
  const radarPresence = Boolean(radar.presence || data.presence);
  const radarVelocity = radar.velocity != null ? Math.abs(Number(radar.velocity)) : 0.24;
  const respiration = radar.respiration != null ? Number(radar.respiration) : null;

  // 1. Temperature checks
  if (temp != null) {
    if (temp > CONFIG.TEMP_HOT_THRESHOLD) {
      alerts.push({
        type: 'ENVIRONMENT',
        severity: temp > (CONFIG.TEMP_HOT_THRESHOLD + 4) ? 'CRITICAL' : 'WARNING',
        message: `High ambient temperature detected (${temp}°C). Exceeds safe threshold (${CONFIG.TEMP_HOT_THRESHOLD}°C). Potential heatwave or HVAC failure.`,
        sensor: 'BME680_TEMP',
        value: temp,
        threshold: CONFIG.TEMP_HOT_THRESHOLD
      });
      flags.push('HIGH_TEMPERATURE');
    } else if (temp < CONFIG.TEMP_COLD_THRESHOLD) {
      alerts.push({
        type: 'ENVIRONMENT',
        severity: temp < (CONFIG.TEMP_COLD_THRESHOLD - 4) ? 'CRITICAL' : 'WARNING',
        message: `Low ambient temperature detected (${temp}°C). Risk of senior hypothermia or room freeze. Threshold: ${CONFIG.TEMP_COLD_THRESHOLD}°C.`,
        sensor: 'BME680_TEMP',
        value: temp,
        threshold: CONFIG.TEMP_COLD_THRESHOLD
      });
      flags.push('LOW_TEMPERATURE');
    }
  }

  // 2. Humidity checks
  if (humidity != null) {
    if (humidity > CONFIG.HUMIDITY_HIGH_THRESHOLD) {
      alerts.push({
        type: 'ENVIRONMENT',
        severity: 'WARNING',
        message: `Elevated relative humidity (${humidity}%). Increased mold risk and dampness hazard.`,
        sensor: 'BME680_HUMIDITY',
        value: humidity,
        threshold: CONFIG.HUMIDITY_HIGH_THRESHOLD
      });
      flags.push('HIGH_HUMIDITY');
    }
  }

  // 3. Combustible & CO Gas checks
  if (gas != null) {
    if (gas >= CONFIG.GAS_CRITICAL_PPM) {
      alerts.push({
        type: 'GAS_HAZARD',
        severity: 'CRITICAL',
        message: `Dangerous gas / carbon monoxide level detected (${gas} ppm). Immediate ventilation and inspection required.`,
        sensor: 'MQ9_GAS',
        value: gas,
        threshold: CONFIG.GAS_CRITICAL_PPM
      });
      flags.push('CRITICAL_GAS');
    } else if (gas >= CONFIG.GAS_WARNING_PPM) {
      alerts.push({
        type: 'GAS_HAZARD',
        severity: 'WARNING',
        message: `Elevated gas concentration detected (${gas} ppm). Above safe baseline.`,
        sensor: 'MQ9_GAS',
        value: gas,
        threshold: CONFIG.GAS_WARNING_PPM
      });
      flags.push('ELEVATED_GAS');
    }
  }

  // 4. Acoustic shock checks
  let isAcousticShock = false;
  if (sound != null && sound >= CONFIG.SOUND_SHOCK_DB) {
    isAcousticShock = true;
    flags.push('ACOUSTIC_SHOCK');
  }

  // 5. Radar velocity sudden change (fall kinematics)
  let isSuddenDescent = false;
  if (radarVelocity >= CONFIG.RADAR_DESCENT_VELOCITY) {
    isSuddenDescent = true;
    flags.push('RAPID_DESCENT');
  }

  // 6. Respiration anomaly checks
  if (respiration != null) {
    if (respiration > CONFIG.RESPIRATION_HIGH_RPM || respiration < CONFIG.RESPIRATION_LOW_RPM) {
      alerts.push({
        type: 'PHYSIOLOGICAL',
        severity: 'WARNING',
        message: `Unusual resting respiration frequency detected (${respiration} RPM). Normal range: ${CONFIG.RESPIRATION_LOW_RPM}–${CONFIG.RESPIRATION_HIGH_RPM} RPM.`,
        sensor: 'RADAR_RESPIRATION',
        value: respiration,
        threshold: respiration < CONFIG.RESPIRATION_LOW_RPM ? CONFIG.RESPIRATION_LOW_RPM : CONFIG.RESPIRATION_HIGH_RPM
      });
      flags.push('ABNORMAL_RESPIRATION');
    }
  }

  // 7. Multi-Sensor Fall / Emergency Corroboration:
  // Corroborates: (Sudden descent radar velocity OR acoustic shock) + (Floor-level presence / PIR trigger)
  let isPossibleFall = false;
  if ((isSuddenDescent || isAcousticShock) && (presence || motion)) {
    isPossibleFall = true;
    alerts.push({
      type: 'FALL_EMERGENCY',
      severity: 'CRITICAL',
      message: 'Possible fall detected. Corroborated multi-vector signature: rapid downward kinetic vector and acoustic impact confirmed.',
      sensor: 'MULTI_CORROBORATION',
      value: radarVelocity,
      threshold: CONFIG.RADAR_DESCENT_VELOCITY
    });
    flags.push('FALL_DETECTED');
  }

  // Determine overall severity
  if (alerts.some(a => a.severity === 'CRITICAL')) {
    highestSeverity = 'CRITICAL';
  } else if (alerts.some(a => a.severity === 'WARNING')) {
    highestSeverity = 'WARNING';
  } else if (temp == null && gas == null && sound == null) {
    highestSeverity = 'UNKNOWN';
  } else {
    highestSeverity = 'NORMAL';
  }

  // Compute confidence score
  let confidence = 98;
  if (highestSeverity === 'CRITICAL') {
    confidence = isPossibleFall ? 87 : 94;
  } else if (highestSeverity === 'WARNING') {
    confidence = 92;
  }

  // Determine human-readable activity pattern and system status
  let activityPattern = 'Normal Activity';
  let systemStatus = 'ONLINE';

  if (isPossibleFall) {
    activityPattern = 'Post-Impact Immobility Detected';
    systemStatus = 'DISPATCH ALERT';
  } else if (isSuddenDescent) {
    activityPattern = 'Abrupt Velocity Vector';
    systemStatus = 'ANALYZING';
  } else if (!motion && presence) {
    activityPattern = 'Sedentary / Resting State';
  } else if (!motion && !presence) {
    activityPattern = 'Room Vacant';
  }

  return {
    status: highestSeverity,
    riskLevel: highestSeverity === 'CRITICAL' ? 'CRITICAL' : highestSeverity === 'WARNING' ? 'ELEVATED' : 'LOW',
    activityPattern,
    systemStatus,
    confidenceScore: confidence,
    flags,
    alerts,
    timestamp: new Date().toISOString()
  };
}

File: backend/app/safety_engine.py
from datetime import datetime, timezone

from app.config import Settings, get_settings


def evaluate_safety_conditions(data: dict, settings: Settings | None = None) -> dict:
    """Classify ambient telemetry. Non-diagnostic; environmental safety only."""
    cfg = settings or get_settings()
    alerts: list[dict] = []
    flags: list[str] = []

    temp = _num(data.get("temperature"))
    humidity = _num(data.get("humidity"))
    gas = _num(data.get("gas"))
    sound = _num(data.get("sound"))
    motion = bool(data.get("motion") or data.get("pir"))
    presence = bool(data.get("presence"))

    radar = data.get("radar") or {}
    radar_velocity = abs(_num(radar.get("velocity")) or 0.24)
    respiration = _num(radar.get("respiration"))
    if respiration is None:
        respiration = _num(data.get("respiration"))

    if temp is not None:
        if temp > cfg.temp_hot_threshold:
            severity = "CRITICAL" if temp > (cfg.temp_hot_threshold + 4) else "WARNING"
            alerts.append(
                {
                    "type": "ENVIRONMENT",
                    "severity": severity,
                    "message": (
                        f"High ambient temperature detected ({temp}°C). "
                        f"Exceeds safe threshold ({cfg.temp_hot_threshold}°C). "
                        "Potential heatwave or HVAC failure."
                    ),
                    "sensor": "BME680_TEMP",
                    "value": temp,
                    "threshold": cfg.temp_hot_threshold,
                }
            )
            flags.append("HIGH_TEMPERATURE")
        elif temp < cfg.temp_cold_threshold:
            severity = "CRITICAL" if temp < (cfg.temp_cold_threshold - 4) else "WARNING"
            alerts.append(
                {
                    "type": "ENVIRONMENT",
                    "severity": severity,
                    "message": (
                        f"Low ambient temperature detected ({temp}°C). "
                        f"Risk of senior hypothermia or room freeze. "
                        f"Threshold: {cfg.temp_cold_threshold}°C."
                    ),
                    "sensor": "BME680_TEMP",
                    "value": temp,
                    "threshold": cfg.temp_cold_threshold,
                }
            )
            flags.append("LOW_TEMPERATURE")

    if humidity is not None and humidity > cfg.humidity_high_threshold:
        alerts.append(
            {
                "type": "ENVIRONMENT",
                "severity": "WARNING",
                "message": (
                    f"Elevated relative humidity ({humidity}%). "
                    "Increased mold risk and dampness hazard."
                ),
                "sensor": "BME680_HUMIDITY",
                "value": humidity,
                "threshold": cfg.humidity_high_threshold,
            }
        )
        flags.append("HIGH_HUMIDITY")

    if gas is not None:
        if gas >= cfg.gas_critical_ppm:
            alerts.append(
                {
                    "type": "GAS_HAZARD",
                    "severity": "CRITICAL",
                    "message": (
                        f"Dangerous gas / carbon monoxide level detected ({gas} ppm). "
                        "Immediate ventilation and inspection required."
                    ),
                    "sensor": "MQ9_GAS",
                    "value": gas,
                    "threshold": cfg.gas_critical_ppm,
                }
            )
            flags.append("CRITICAL_GAS")
        elif gas >= cfg.gas_warning_ppm:
            alerts.append(
                {
                    "type": "GAS_HAZARD",
                    "severity": "WARNING",
                    "message": f"Elevated gas concentration detected ({gas} ppm). Above safe baseline.",
                    "sensor": "MQ9_GAS",
                    "value": gas,
                    "threshold": cfg.gas_warning_ppm,
                }
            )
            flags.append("ELEVATED_GAS")

    is_acoustic_shock = sound is not None and sound >= cfg.sound_shock_db
    if is_acoustic_shock:
        flags.append("ACOUSTIC_SHOCK")

    is_sudden_descent = radar_velocity >= cfg.radar_descent_velocity
    if is_sudden_descent:
        flags.append("RAPID_DESCENT")

    if respiration is not None and (
        respiration > cfg.respiration_high_rpm or respiration < cfg.respiration_low_rpm
    ):
        threshold = (
            cfg.respiration_low_rpm
            if respiration < cfg.respiration_low_rpm
            else cfg.respiration_high_rpm
        )
        alerts.append(
            {
                "type": "PHYSIOLOGICAL",
                "severity": "WARNING",
                "message": (
                    f"Unusual resting respiration frequency detected ({respiration} RPM). "
                    f"Normal range: {cfg.respiration_low_rpm}–{cfg.respiration_high_rpm} RPM."
                ),
                "sensor": "RADAR_RESPIRATION",
                "value": respiration,
                "threshold": threshold,
            }
        )
        flags.append("ABNORMAL_RESPIRATION")

    is_possible_fall = (is_sudden_descent or is_acoustic_shock) and (presence or motion)
    if is_possible_fall:
        alerts.append(
            {
                "type": "FALL_EMERGENCY",
                "severity": "CRITICAL",
                "message": (
                    "Possible fall detected. Corroborated multi-vector signature: "
                    "rapid downward kinetic vector and acoustic impact confirmed."
                ),
                "sensor": "MULTI_CORROBORATION",
                "value": radar_velocity,
                "threshold": cfg.radar_descent_velocity,
            }
        )
        flags.append("FALL_DETECTED")

    if any(a["severity"] == "CRITICAL" for a in alerts):
        highest = "CRITICAL"
    elif any(a["severity"] == "WARNING" for a in alerts):
        highest = "WARNING"
    elif temp is None and gas is None and sound is None:
        highest = "UNKNOWN"
    else:
        highest = "NORMAL"

    confidence = 98
    if highest == "CRITICAL":
        confidence = 87 if is_possible_fall else 94
    elif highest == "WARNING":
        confidence = 92

    activity_pattern = "Normal Activity"
    system_status = "ONLINE"
    if is_possible_fall:
        activity_pattern = "Post-Impact Immobility Detected"
        system_status = "DISPATCH ALERT"
    elif is_sudden_descent:
        activity_pattern = "Abrupt Velocity Vector"
        system_status = "ANALYZING"
    elif not motion and presence:
        activity_pattern = "Sedentary / Resting State"
    elif not motion and not presence:
        activity_pattern = "Room Vacant"

    return {
        "status": highest,
        "riskLevel": "CRITICAL" if highest == "CRITICAL" else "ELEVATED" if highest == "WARNING" else "LOW",
        "activityPattern": activity_pattern,
        "systemStatus": system_status,
        "confidenceScore": confidence,
        "flags": flags,
        "alerts": alerts,
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }


def _num(value):
    if value is None or value == "":
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None
